### DISCUSSION TITLE
Retry a transaction on SQLITE_BUSY

### DISCUSSION THREAD

**Discussion by joggienl** [directus/directus#23236] [Feature Requests]
### Summary

In case of using SQLite as database and set the SQLite database to be in WAL mode (`PRAGMA journal_mode = WAL;`) it is almost inevitable to not get an `SQLITE_BUSY` error at some point. Especially if you manage to derive from the default knex.js pool configuration for SQLite with min and max being set to just `1`.

There are several ways to prevent and or handle those errors:

- Use transactions (already being done)
- Set `busy_timeout`
- Retry in case of `SQLITE_BUSY`

### Motivation

Changing `api/src/utils/transaction.ts` is pretty straightforward. There is already working code there that retries in case of `COCKROACH_RETRY_ERROR_CODE`.

By extending this a bit to also retry in case of the `SQLITE_BUSY` error we hugely improve the support for concurrent SQLite usage.

The `SQLITE_BUSY` error does happen when you add multiple relations. This is because the studio will send multiple calls to `/relations` concurrent. the `alterTable` and more specifically `builder.reference()` and `builder.foreign()` don't trigger SQLite in write mode causing the error to occur.

### Detailed Design

Refactor `api/src/utils/transaction.ts` to also retry if there is an `SQLITE_BUSY` error:

```js
const COCKROACH_RETRY_ERROR_CODE = '40001';
const SQLITE_BUSY_ERROR_CODE = 'SQLITE_BUSY';

if (
  (client !== 'cockroachdb' && client !== 'sqlite3') ||
  (error?.code !== COCKROACH_RETRY_ERROR_CODE && error?.code !== SQLITE_BUSY_ERROR_CODE)
) throw error;
```

### Must Have:
- Retry the transaction if SQLITE_BUSY happens

### Should Have:
- .

### Could Have:
- .

### Won't Have:
- how to on setting up WAL mode and others to enable concurrent SQLite database

### Drawbacks

I can't think of any drawbacks but am happy to add if anyone sees any!

### Alternatives

As far as I know I do not see another way to add a retry decent.

### Adoption Strategy

This change should NOT affect other current processes.

**Comment by joggienl** [comment]
Happy to add a PR if accepted, tested this locally already!

**Reply by rijkvanzanten** [reply]
Sounds good to me! Looks like a relatively small tweak that would have a huge quality of life improvement 🙂 Go for it! 🚀

**Reply by joggienl** [reply]
thanks! I've opened https://github.com/directus/directus/pull/23243

**Reply by rijkvanzanten** [reply]
Merged! 🥳 Thanks!

**Reply by joggienl** [reply]
Love it, thanks for merging, great minor improvement!
